util/file_reader_writer.cc

```diff
@@ -59,81 +59,116 @@ Status WritableFileWriter::Append(const Slice& data) {
     TEST_SYNC_POINT("WritableFileWriter::Append:BeforePrepareWrite");
     writable_file_->PrepareWrite(static_cast<size_t>(GetFileSize()), left);
   }
-  // if there is no space in the cache, then flush
-  if (cursize_ + left > capacity_) {
-    s = Flush();
-    if (!s.ok()) {
-      return s;
+
+  // Flush only when I/O is buffered
+  if (use_os_buffer_ &&
+    (buf_.GetCapacity() - buf_.GetCurrentSize()) < left) {
+    if (buf_.GetCurrentSize() > 0) {
+      s = Flush();
+      if (!s.ok()) {
+        return s;
+      }
     }
-    // Increase the buffer size, but capped at 1MB
-    if (capacity_ < (1 << 20)) {
-      capacity_ *= 2;
-      buf_.reset(new char[capacity_]);
+
+    if (buf_.GetCapacity() < (1 << 20)) {
+      size_t desiredCapacity = buf_.GetCapacity() * 2;
+      desiredCapacity = std::min(desiredCapacity, size_t(1 << 20));
```

Comment:
Can we do it as a follow-up? I don't want to mix behavior change with this restructuring.